src/ClientHandle.cpp

```diff
@@ -1395,187 +1395,166 @@ void cClientHandle::FinishDigAnimation()
 
 
 
-void cClientHandle::HandleRightClick(int a_BlockX, int a_BlockY, int a_BlockZ, eBlockFace a_BlockFace, int a_CursorX, int a_CursorY, int a_CursorZ, const cItem & a_HeldItem)
-{
-	// TODO: Rewrite this function
-
+void cClientHandle::HandleRightClick(int a_BlockX, int a_BlockY, int a_BlockZ, eBlockFace a_BlockFace, int a_CursorX, int a_CursorY, int a_CursorZ, int a_Hand)
+{
+	// This function handles three actions:
+	// (1) Place a block
+	// (2) "Use" a block: Interactive with the block, like opening a chest/crafting table/furnace
+	// (3) Use the held item targeting a block: e.g. farming
+	// (4) Use the held item without targeting a block: eating, drinking, charging a bow, using buckets
+	//
+	// As for 1.12.2, This function has two use cases:
+	// * Calling with parameters -1, 255, -1, BLOCK_FACE_NONE(-1), 0, 0, 0, Hand. --(4)
+	// * Calling with valid parameters.                                           --(1) (2) or (3)
+	//   Sneaking player will prioritize placement over using block, otherwise the other way round.
+	//
+	// In version 1.8.x, these two cases shareing the same packet id. In version >= 1.9, there is a new packet id "Use Item".
+	// But we still call this function in its handler for backward compatibility.
+    //
+	// Frozen player can do nothing.
```

Comment:
I would expect, that this statement means, that we do a fast bail out for frozen players. Or does this function needs to handle these players?

Reply:
The server needs to send back an update of the block otherwise the client would assume placing the block successfully.